src/ResourceManager/Sql/Commands.Sql/Common/AzureSqlCmdletBaseBase.cs

```diff
@@ -0,0 +1,169 @@
+﻿
+
+// ----------------------------------------------------------------------------------
+//
+// Copyright Microsoft Corporation
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Azure.Commands.Common.Authentication.Models;
+using Microsoft.Azure.Commands.ResourceManager.Common;
+using Microsoft.Azure.Commands.Sql.Services;
+using System.Management.Automation;
+using Microsoft.Azure.Commands.Common.Authentication.Abstractions;
+using System.Collections;
+
+namespace Microsoft.Azure.Commands.Sql.Common
+{
+    /// <summary>
+    /// The base class for all Azure Sql cmdlets
+    /// </summary>
+    public abstract class AzureSqlCmdletBaseBase<M, A> : AzureRMCmdlet 
+    {
+        /// <summary>
+        /// Stores the per request session Id for all request made in this cmdlet call.
+        /// </summary>
+        protected string clientRequestId { get; set; }
+
+        /// <summary>
+        /// Default constructor.  Generates a request ID
+        /// </summary>
+        internal AzureSqlCmdletBaseBase()
+        {
+        }
+
+        protected virtual string GetResourceId(M model)
```

Comment:
@shahabhijeet  This would be overridden per cmdlet type (Database, Server, DataMasking, ...) and as such the type M will be well known and there shouldn't be the complex logic I have.